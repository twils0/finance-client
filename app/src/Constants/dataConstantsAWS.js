// eslint-disable-next-line
export const actionTypes = {
  SET_AWS_STATUS: 'DATA_AUTH_AWS_STATUS',
  SET_AWS_USER: 'DATA_AUTH_SET_AWS_USER',
  RESET_AWS_STATE: 'DATA_AUTH_RESET_AWS_STATE',
};
