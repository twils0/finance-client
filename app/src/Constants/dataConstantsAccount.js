export const actionTypes = {
  SET_ACCOUNT_STATUS: 'DATA_ACCOUNT_SET_ACCOUNT_STATUS',
  SET_FIELD: 'DATA_ACCOUNT_SET_FIELD',
  SET_FIELDS: 'DATA_ACCOUNT_SET_FIELDS',
  RESET_ACCOUNT_STATE: 'DATA_ACCOUNT_RESET_ACCOUNT_STATE',
};

export const statusNames = {
  AWS_FIELDS: 'accountStatusAWSFields',
  STRIPE_FIELDS: 'accountStatusStripeFields',
  UPDATE_STRIPE_FIELDS_TOKEN: 'accountStatusUpdateStripeFieldsToken',
  UPDATE_STRIPE_FIELDS_REQUEST: 'accountStatusUpdateStripeFieldsRequest',
  UPDATE_DB_FIELDS: 'accountStatusUpdateDBFields',
  UPDATE_AWS_FIELDS: 'accountStatusUpdateAWSFields',
};

export const fieldNames = {
  NAME: 'accountName',
  EMAIL: 'accountEmail',
  EMAIL_ADDITIONAL: 'accountEmailAdditional',
  PHONE: 'accountPhone',
  NAME_ON_CARD: 'accountNameOnCard',
  PROMO_CODE: 'accountPromoCode',
  PROMO_CODE_VALID: 'accountPromoCodeValid',
  CUSTOMER: 'accountCustomer',
};
