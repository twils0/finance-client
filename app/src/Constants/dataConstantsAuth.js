export const actionTypes = {
  SET_AUTHENTICATED: 'DATA_AUTH_SET_AUTHENTICATED',
  SET_REDIRECT_URL: 'DATA_AUTH_SET_REDIRECT_URL',
  SET_AUTH_STATUS: 'DATA_AUTH_SET_AUTH_STATUS',
  SET_CODE_TYPE: 'DATA_AUTH_SET_CODE_TYPE',
  RESET_AUTH_STATE: 'DATA_AUTH_RESET_AUTH_STATE',
};

export const statusNames = {
  LOGIN: 'authStatusLogin',
  SIGN_OUT_DEVICES: 'authStatusSignOutDevices',
  LOGIN_MFA: 'authStatusLoginMFA',
  SIGN_UP: 'authStatusSignUp',
  VERIFY_PHONE: 'authStatusVerifyPhone',
  VERIFY_PHONE_CODE: 'authStatusVerifyPhoneCode',
  VERIFY_EMAIL: 'authStatusVerifyEmail',
  VERIFY_EMAIL_LINK: 'authStatusVerifyEmailLink',
  FORGOT_PASSWORD: 'authStatusForgotPassword',
  RESET_PASSWORD: 'authStatusResetPassword',
  CHANGE_PASSWORD: 'authStatusChangePassword',
  DELETE_ACCOUNT: 'authStatusDeleteAccount',
  LOGOUT: 'authStatusLogout',
};

export const codeTypeNames = {
  VERIFY_PHONE: 'verifyPhone',
  VERIFY_EMAIL: 'verifyEmail',
  VERIFY_EMAIL_ADDITIONAL: 'verifyEmailAdditional',
};
