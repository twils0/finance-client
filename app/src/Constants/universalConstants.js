import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
