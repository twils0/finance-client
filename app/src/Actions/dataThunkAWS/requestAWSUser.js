import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestLogout from '../dataThunkAuth/requestLogout';

import { demo } from '../../../../mode.config.json';

const requestAWSUser = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.aws.status !== requestStatusTypes.LOADING) {
    dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

    let user = null;

    try {
      if (!demo) {
        user = await Auth.currentUserPoolUser();
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));

      if (
        (typeof error === 'string'
          && (error === 'not authenticated' || error === 'No current user in userPool'))
        || (typeof error === 'object' && error && error.code === 'NotAuthorizedException')
      ) {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestAWSUser',
        });
      }

      return Promise.reject(error);
    }

    if (!demo) {
      dispatch(setAWSUser({ user }));
    }
    dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

    return { user };
  }

  return null;
};

export default requestAWSUser;
