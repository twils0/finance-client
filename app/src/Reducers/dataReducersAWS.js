import { combineReducers } from 'redux';

import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes } from '../Constants/dataConstantsAWS';

export const initialState = {
  status: requestStatusTypes.IDLE,
  user: null,
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_STATUS:
      return action.payload.status;
    case actionTypes.RESET_AWS_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const user = (state = initialState.user, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_USER:
      return action.payload.user;
    case actionTypes.RESET_AWS_STATE:
      return initialState.user;
    default:
      return state;
  }
};

const dataReducersAWS = combineReducers({
  status,
  user,
});

export default dataReducersAWS;
