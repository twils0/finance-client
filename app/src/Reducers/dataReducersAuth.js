import { combineReducers } from 'redux';

import { requestStatusTypes, pathNames } from '../Constants/universalConstants';
import { actionTypes, statusNames, codeTypeNames } from '../Constants/dataConstantsAuth';

export const initialState = {
  authenticated: false,
  redirectURL: pathNames.WATCHLIST,
  status: {
    [statusNames.LOGIN]: {
      id: statusNames.LOGIN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_OUT_DEVICES]: {
      id: statusNames.SIGN_OUT_DEVICES,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGIN_MFA]: {
      id: statusNames.LOGIN_MFA,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_UP]: {
      id: statusNames.SIGN_UP,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE]: {
      id: statusNames.VERIFY_PHONE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE_CODE]: {
      id: statusNames.VERIFY_PHONE_CODE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL]: {
      id: statusNames.VERIFY_EMAIL,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL_LINK]: {
      id: statusNames.VERIFY_EMAIL_LINK,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.FORGOT_PASSWORD]: {
      id: statusNames.FORGOT_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.RESET_PASSWORD]: {
      id: statusNames.RESET_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.CHANGE_PASSWORD]: {
      id: statusNames.CHANGE_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGOUT]: {
      id: statusNames.LOGOUT,
      status: requestStatusTypes.SUCCESS,
    },
    [statusNames.DELETE_ACCOUNT]: {
      id: statusNames.DELETE_ACCOUNT,
      status: requestStatusTypes.SUCCESS,
    },
  },
  codeTypes: {
    [codeTypeNames.VERIFY_PHONE]: {
      id: codeTypeNames.VERIFY_PHONE,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL]: {
      id: codeTypeNames.VERIFY_EMAIL,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL_ADDITIONAL]: {
      id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
      needed: true,
    },
  },
};

export const authenticated = (state = initialState.authenticated, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTHENTICATED:
      return action.payload.authenticated;
    case actionTypes.RESET_AUTH_STATE:
      return initialState.authenticated;
    default:
      return state;
  }
};

export const redirectURL = (state = initialState.redirectURL, action) => {
  switch (action.type) {
    case actionTypes.SET_REDIRECT_URL:
      return action.payload.redirectURL;
    default:
      return state;
  }
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTH_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const codeTypes = (state = initialState.codeTypes, action) => {
  switch (action.type) {
    case actionTypes.SET_CODE_TYPE:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.codeTypes;
    default:
      return state;
  }
};

const dataReducersAuth = combineReducers({
  authenticated,
  redirectURL,
  status,
  codeTypes,
});

export default dataReducersAuth;
