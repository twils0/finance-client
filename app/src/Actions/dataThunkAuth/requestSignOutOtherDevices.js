import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus } from '../dataActionsAuth';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';

const requestSignOutOtherDevices = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'email')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'password')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'remembered')) {
    throw new Error(`Please enter a value for the 'remembered' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    let state = getState();

    if (
      state.data.auth.status[statusNames.SIGN_OUT_DEVICES].status !== requestStatusTypes.LOADING
    ) {
      dispatch(setAuthStatus({
        id: statusNames.SIGN_OUT_DEVICES,
        status: requestStatusTypes.LOADING,
      }));

      let { user } = state.data.aws;

      try {
        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await new Promise((resolve, reject) => {
          user.setDeviceStatusRemembered({
            onSuccess: () => {
              resolve();
            },
            onFailure: (error) => {
              reject(error);
            },
          });
        });

        await new Promise((resolve, reject) => {
          user.globalSignOut({
            onSuccess: () => {
              resolve();
            },
            onFailure: (error) => {
              reject(error);
            },
          });
        });

        dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

        user = await Auth.signIn(payload.email, payload.password);

        dispatch(setAWSUser({ user }));
        dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

        if (!payload.remembered) {
          await new Promise((resolve, reject) => {
            user.setDeviceStatusNotRemembered({
              onSuccess: () => {
                resolve();
              },
              onFailure: (error) => {
                reject(error);
              },
            });
          });
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);
        state = getState();

        dispatch(setAuthStatus({
          id: statusNames.SIGN_OUT_DEVICES,
          status: requestStatusTypes.ERROR,
        }));
        if (state.data.aws.status === requestStatusTypes.LOADING) {
          dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));
        }

        raven.captureException(error, {
          logger: 'requestSignOutOtherDevices',
        });

        return Promise.reject(error);
      }

      dispatch(setAuthStatus({
        id: statusNames.SIGN_OUT_DEVICES,
        status: requestStatusTypes.SUCCESS,
      }));

      return user;
    }

    return null;
  };
};

export default requestSignOutOtherDevices;
