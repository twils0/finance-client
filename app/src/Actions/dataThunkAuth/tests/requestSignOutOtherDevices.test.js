import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Auth } from 'aws-amplify';

import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { initialState as initialStateAWS } from '../../../Reducers/dataReducersAWS';
import { requestStatusTypes } from '../../../Constants/universalConstants';
import { actionTypes as actionTypesAuth, statusNames } from '../../../Constants/dataConstantsAuth';
import { actionTypes as actionTypesAWS } from '../../../Constants/dataConstantsAWS';
import requestAWSUser from '../../dataThunkAWS/requestAWSUser';
import requestSignOutOtherDevices from '../requestSignOutOtherDevices';

jest.mock('aws-amplify', () => ({
  Auth: {
    signIn: jest.fn(),
  },
}));
jest.mock('../../dataThunkAWS/requestAWSUser', () => jest.fn());

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const userObject = { user: { test: 'testUser' } };
const signOutPayload = { email: 'testEmail', password: 'testPassword1!', remembered: true };

describe('dataThunkAuth', () => {
  describe('requestSignOutOtherDevices', () => {
    afterEach(() => {
      requestAWSUser.mockReset();
      if (userObject.user.setDeviceStatusRemembered) {
        delete userObject.user.setDeviceStatusRemembered;
      }
      if (userObject.user.globalSignOut) {
        delete userObject.user.globalSignOut;
      }
      Auth.signIn.mockReset();
      if (userObject.user.setDeviceStatusNotRemembered) {
        delete userObject.user.setDeviceStatusNotRemembered;
      }
    });

    it("fails and throws an error when missing 'email' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const emptyPayload = {};

      try {
        await store.dispatch(requestSignOutOtherDevices(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
    });

    it("fails and throws an error when missing 'password' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const wrongPayload = { email: 'testEmail' };

      try {
        await store.dispatch(requestSignOutOtherDevices(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
    });

    it("fails and throws an error when missing 'remembered' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const wrongPayload = { email: 'testEmail', password: 'testPassword1!' };

      try {
        await store.dispatch(requestSignOutOtherDevices(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(requestAWSUser).not.toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
    });

    it("calls requestAWSUser when missing 'user' object in state", async () => {
      const store = mockStore({ data: { aws: initialStateAWS, auth: initialStateAuth } });

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());

      requestAWSUser.mockReturnValue(() => Promise.resolve(userObject));
      Auth.signIn.mockReturnValue(Promise.resolve(userObject.user));

      await store.dispatch(requestSignOutOtherDevices(signOutPayload));

      expect(requestAWSUser).toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).toBeCalledWith(signOutPayload.email, signOutPayload.password);
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });

    it('fails and returns promise reject when requestAWSUser throws an error', async () => {
      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());

      const store = mockStore({ data: { aws: initialStateAWS, auth: initialStateAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      requestAWSUser.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestSignOutOtherDevices(signOutPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).not.toBeCalled();
      expect(userObject.user.globalSignOut).not.toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });

    it('fails and returns promise reject when user.setDeviceStatusRemembered throws an error', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      const error = { code: 'testError', message: 'testMessage' };

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onFailure }) => onFailure(error));
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      try {
        await store.dispatch(requestSignOutOtherDevices(signOutPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).not.toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });

    it('fails and returns promise reject when user.globalSignOut throws an error', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      const error = { code: 'testError', message: 'testMessage' };

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onFailure }) => onFailure(error));
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      try {
        await store.dispatch(requestSignOutOtherDevices(signOutPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).not.toBeCalled();
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });

    it('fails and returns promise reject when Auth.signIn throws an error', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());
      stateBeforeAWS.user = userObject.user;
      stateBeforeAWS.status = requestStatusTypes.LOADING;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.ERROR,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.reject(error));

      try {
        await store.dispatch(requestSignOutOtherDevices(signOutPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).toBeCalledWith(signOutPayload.email, signOutPayload.password);
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });

    it('fails and returns promise reject when user.setDeviceStatusNotRemembered throws an error when remembered is false', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      const error = { code: 'testError', message: 'testMessage' };

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onFailure }) => onFailure(error));
      stateBeforeAWS.user = userObject.user;

      const signOutNotRememberedPayload = {
        email: 'testEmail',
        password: 'testPassword1!',
        remembered: false,
      };
      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user: userObject.user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(userObject.user));

      try {
        await store.dispatch(requestSignOutOtherDevices(signOutNotRememberedPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).toBeCalledWith(signOutPayload.email, signOutPayload.password);
      expect(userObject.user.setDeviceStatusNotRemembered).toBeCalled();
    });

    it('creates the correct actions with the correct payloads when remembered is false', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());
      stateBeforeAWS.user = userObject.user;

      const signOutNotRememberedPayload = {
        email: 'testEmail',
        password: 'testPassword1!',
        remembered: false,
      };
      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user: userObject.user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(userObject.user));

      await store.dispatch(requestSignOutOtherDevices(signOutNotRememberedPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).toBeCalledWith(signOutPayload.email, signOutPayload.password);
      expect(userObject.user.setDeviceStatusNotRemembered).toBeCalled();
    });

    it('creates the correct actions with the correct payloads', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));

      userObject.user.setDeviceStatusRemembered = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.globalSignOut = jest.fn(({ onSuccess }) => onSuccess());
      userObject.user.setDeviceStatusNotRemembered = jest.fn(({ onSuccess }) => onSuccess());
      stateBeforeAWS.user = userObject.user;

      const store = mockStore({ data: { aws: stateBeforeAWS, auth: initialStateAuth } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user: userObject.user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_OUT_DEVICES,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(userObject.user));

      await store.dispatch(requestSignOutOtherDevices(signOutPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(userObject.user.setDeviceStatusRemembered).toBeCalled();
      expect(userObject.user.globalSignOut).toBeCalled();
      expect(Auth.signIn).toBeCalledWith(signOutPayload.email, signOutPayload.password);
      expect(userObject.user.setDeviceStatusNotRemembered).not.toBeCalled();
    });
  });
});
