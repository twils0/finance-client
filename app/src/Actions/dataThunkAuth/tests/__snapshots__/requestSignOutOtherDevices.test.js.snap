// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataThunkAuth requestSignOutOtherDevices fails and throws an error when missing 'email' key in payload 1`] = `[Error: Please enter a value for the 'password' key - {}]`;

exports[`dataThunkAuth requestSignOutOtherDevices fails and throws an error when missing 'password' key in payload 1`] = `[Error: Please enter a value for the 'password' key - {"email":"testEmail"}]`;

exports[`dataThunkAuth requestSignOutOtherDevices fails and throws an error when missing 'remembered' key in payload 1`] = `[Error: Please enter a value for the 'remembered' key - {"email":"testEmail","password":"testPassword1!"}]`;
